### DISCUSSION TITLE
Firestore?

### DISCUSSION THREAD

**Discussion by wiverson** [capawesome-team/capacitor-firebase#228] [Q&A]
Ok, perhaps this is obvious, but how should I use Cloud Firestore w/Capacitor?

The only package I can find on npm w/both capacitor and firestore tags is this one - is this the only option?

https://github.com/ProteanSoftware/capacitor-firestore

Or is there a capawesome plugin, or...?

**Comment by robingenz** [comment]
> Ok, perhaps this is obvious, but how should I use Cloud Firestore w/Capacitor?

Short answer: Use the Firebase JS SDK.

You can also use the Firebase JS SDK with the Capacitor Firebase Authentication plugin by following this guide: https://github.com/capawesome-team/capacitor-firebase/blob/main/packages/authentication/docs/firebase-js-sdk.md

Related to https://github.com/capawesome-team/capacitor-firebase/issues/41

**Reply by JTCorrin** [reply]
@robingenz I'm not able to use firestore in another plugin after authenticating via this plugin. Am I missing something? I've followed the instructions you've posted above. Here is my login snippet:

```
if (Capacitor.isNativePlatform()) {
    await firebase.auth().setPersistence(firebase.auth.Auth.Persistence.LOCAL)
}

// Sign in on native layer
const result = await FirebaseAuthentication.signInWithEmailAndPassword({
    email,
    password
})

// Use response to sign in on weblayer
await firebase.auth().signInWithCredential(result.credential)
```

From what I can see this logs in correctly.

I'm developing another capacitor plugin that uses the native firestore libraries. When I call a firestore.get() request for example I get

```
com.google.firebase.firestore.FirebaseFirestoreException: PERMISSION_DENIED: Missing or insufficient permissions.
```

So far I've only tested on Android. Do I need to do anything extra to get the native auth state to be available to other plugins?

Thanks!

**Reply by robingenz** [reply]
@JTCorrin You only need to do this if you want to use the Firebase JS SDK. If you want to use the Firebase Android SDK or Firebase iOS SDK for Firestore, then you don't need the Firebase JS SDK and can call the methods of this plugin as usual:

```ts
import { FirebaseAuthentication } from '@capacitor-firebase/authentication';

const signInWithEmailAndPassword = async () => {
  const result = await FirebaseAuthentication.signInWithEmailAndPassword({
    email: 'mail@example.com',
    password: '1234',
  });
  return result.user;
};
```

If this does not work, please create a new issue with a Minimal, Reproducible Example (see [How to create a Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example)) in a public GitHub repository so I can debug it.

**Reply by JTCorrin** [reply]
Thanks for the quick response, @robingenz . 

I actually just got something working like this 

```
if (Capacitor.isNativePlatform()) {
		await firebase.auth().setPersistence(firebase.auth.Auth.Persistence.LOCAL)
	}

	// Sign in on native layer
	const result = await FirebaseAuthentication.signInWithEmailAndPassword({
		email,
		password
	})

	if (result.credential !== null) {
		const creds = firebase.auth.EmailAuthProvider.credential(email, password)
		await firebase.auth().signInWithCredential(creds)
	}
```
Although, the second login seems redundant.

The reason I am using the Firebase js SDK is because my webview layer is littered with firestore calls. I'm adding firestore natively as well to get around one specific issue.

Maybe I should not take this approach and use a native http call to my backend to do this specific task. 

Any opinion on that?

Thanks again!

**Reply by robingenz** [reply]
It depends on your specific use case, unfortunately I can't give you much advice without more information. 

> Maybe I should not take this approach and use a native http call to my backend to do this specific task.

That sounds the easiest.

If you have more questions, please create a new discussion as this is off-topic.
